File: 001_Simulation_und_Schaltungsentwurf/005_self_tuned_filter/redpitaya_scpi/diodes_data_meas.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Messautomatisierung - RedPitaya STEMlab SCPI

@author: T. Ziemann, M. Meiners
"""

# %% Init
import time
import datetime
import numpy as np
import pandas as pd
# import matplotlib.pyplot as plt
import redpitaya_scpi as scpi

LABDESK = {
    "ELIE1": "192.168.111.181",
    "ELIE2": "192.168.111.182",
    "ELIE3": "192.168.111.183",
    "ELIE4": "192.168.111.184",
    "ELIE5": "192.168.111.185",
    "ELIE6": "192.168.111.186"
}

IP = LABDESK["ELIE1"]  # IP des Laborplatzes
rp = scpi.scpi(IP)

# size in samples 16 Bit
DATA_SIZE = 1024 * 16          # ((1024 * 1024 * 128) / 2) ## for 128 MB ##
READ_DATA_SIZE = 1024 * 16     # (1024 * 256)              ## for 128 MB ##

dec = 16
trig_lvl = 0.4

# %% Messung / Datenverarbeitung
TSTAMP = datetime.datetime.now()
DEVICE = {
    "d1": "1N4001",
    "d2": "1N4148",
    "npn": "2N3904",
    "pnp": "2N3906",
    "nmos": "BS170",
    "pmos": "BS250"
}


# Init Dataframes
DF_IN1 = pd.DataFrame()
DF_IN2 = pd.DataFrame()


# Konfiguration des Signalgenerators
func = 'TRIANGLE'  # Signalform (SINE, SQUARE, TRIANGLE)
ampl = 0.5  # Amplitude (-1 V ... 1 V)
freq = 1000  # Frequenz (0 Hz ... 62.5e6 Hz)
offset = 0.5  # Offset (-1 V ... 1 V)
t = np.linspace(0, 2.097e-3, 16384)  # Skalierung der Zeitachse (Sampling)

rp.tx_txt('ANALOG:RST ')  # Set analog outputs to 0V
rp.tx_txt('GEN:RST')  # Signal Generator reset
rp.tx_txt('PHAS:ALIGN')  # Synchronize
rp.tx_txt('SOUR1:TRig:INT')
rp.tx_txt('SOUR1:FUNC ' + str(func))  # Wellenform
rp.tx_txt('SOUR1:VOLT ' + str(ampl))  # Amplitude
rp.tx_txt('SOUR1:VOLT:OFFS ' + str(offset))  # Offset
rp.tx_txt('SOUR1:PHAS ' + str(offset))  # Offset
rp.tx_txt('SOUR1:FREQ:FIX:Direct ' + str(freq))  # Frequenz
# rp.tx_txt('SOUR1:FREQ:FIX ' + str(freq))  # Frequenz

# Enable output
rp.tx_txt('OUTPUT1:STATE ON')
rp.tx_txt('SOUR1:TRig:INT')

time.sleep(1)  # in Sekunden

# print("Start program")

# ACQUISITION
rp.tx_txt('ACQ:RST')  # Input reset

# Get memory region (DMM)
start_address = int(rp.txrx_txt('ACQ:AXI:START?'))
size = int(rp.txrx_txt('ACQ:AXI:SIZE?'))
start_address2 = round(start_address + size/2)


rp.tx_txt(f"ACQ:AXI:DEC {dec}")  # Decimation
# rp.tx_txt(f"ACQ:DEC {dec}")  # Decimation (1, 8, 16, 64, 1024, 8192)

rp.tx_txt('ACQ:AXI:DATA:Units VOLTS')  # Set units
# rp.tx_txt('ACQ:DATA:Units VOLTS')

rp.tx_txt(f"ACQ:AXI:SOUR1:Trig:Dly {DATA_SIZE}")  # Trigger delay ch1
rp.tx_txt(f"ACQ:AXI:SOUR2:Trig:Dly {DATA_SIZE}")  # Trigger delay ch2
# rp.tx_txt(f"ACQ:TRig:DLY {DATA_SIZE}")  # Delay

# Set-up channel 1 and 2 buffers to each work with half of available memory space.
rp.tx_txt(f"ACQ:AXI:SOUR1:SET:Buffer {start_address},{size/2}")
rp.tx_txt(f"ACQ:AXI:SOUR2:SET:Buffer {start_address2},{size/2}")

rp.tx_txt('ACQ:AXI:SOUR1:ENable ON')  # Enable DMM ch1
rp.tx_txt('ACQ:AXI:SOUR2:ENable ON')  # Enable DMM ch2

rp.tx_txt(f"ACQ:TRig:LEV {trig_lvl}")  # Trigger level


rp.tx_txt('ACQ:START')  # Start aquisition
rp.tx_txt('ACQ:TRig NOW')  # Trigger manually

# print("Waiting for trigger\n")
while 1:
    rp.tx_txt('ACQ:TRig:STAT?')
    if rp.rx_txt() == 'TD':
        break

# while 1:
#    rp.tx_txt('ACQ:TRig:FILL?')
#    if rp.rx_txt() == '1':
#        break

# wait for fill adc buffer DMM
while 1:
    rp.tx_txt('ACQ:AXI:SOUR1:TRig:FILL?')
    if rp.rx_txt() == '1':
        print('DMA buffer full\n')
        break

# Data Acquisition
for meas in range(0, 1):

    # Input IN1
    # time.sleep(0.1)  # in seconds
    # rp.tx_txt('ACQ:SOUR1:DATA?')  # Readout buffer IN1
    posIN1 = int(rp.txrx_txt('ACQ:AXI:SOUR1:Trig:Pos?'))
    rp.tx_txt(f"ACQ:AXI:SOUR1:DATA:Start:N? {posIN1},{READ_DATA_SIZE}")
    IN1str = rp.rx_txt()
    IN1str = IN1str.strip('{}\n\r').replace("  ", "").split(',')
    DF_IN1[str(meas)] = np.array(list(map(float, IN1str)))

    # Get write pointer at trigger location

    # Input IN2
    # time.sleep(0.1)  # in seconds
    # rp.tx_txt('ACQ:SOUR2:DATA?')  # Readout buffer IN2
    posIN2 = int(rp.txrx_txt('ACQ:AXI:SOUR2:Trig:Pos?'))
    rp.tx_txt(f"ACQ:AXI:SOUR2:DATA:Start:N? {posIN2},{READ_DATA_SIZE}")
    IN2str = rp.rx_txt()
    IN2str = IN2str.strip('{}\n\r').replace("  ", "").split(',')
    DF_IN2[str(meas)] = np.array(list(map(float, IN2str)))


# Stop Acquisition
rp.tx_txt('ACQ:STOP')

# Stopp des Generators
rp.tx_txt('OUTPUT1:STATE OFF')

# %% Daten als Datei speichern
Data_IN1 = 'data/IN1_' + DEVICE["d1"] + \
    "_ELIE1" + str(TSTAMP.strftime('_%Y-%m-%d_%H-%M'))
Data_IN2 = 'data/IN2_' + DEVICE["d1"] + \
    "_ELIE1" + str(TSTAMP.strftime('_%Y-%m-%d_%H-%M'))


# %% Speichern als CSV/TSV (comma/tab-seperated-values)
DF_IN1.to_csv(Data_IN1 + '.csv', index=False)
DF_IN2.to_csv(Data_IN2 + '.csv', index=False)

# %% Speichern als Excel-Worksheet
# with pd.ExcelWriter(Data_IN + '.xlsx') as writer:
#     DF_IN1.to_excel(writer, sheet_name='IN1', index=False)
#     DF_IN2.to_excel(writer, sheet_name='IN2', index=False)

# DF_IN1.to_excel(Data_IN1 + '.xlsx', index=False)
# DF_IN2.to_excel(Data_IN2 + '.xlsx', index=False)

# %% Speichern als Apache parquet
# DF_IN1.to_parquet(Data_IN1 + ".parquet", index=False)
# DF_IN2.to_parquet(Data_IN2 + ".parquet", index=False)

File: 001_Simulation_und_Schaltungsentwurf/005_self_tuned_filter/redpitaya_scpi/diodes_data_meas.m
%% STEMlab measurements for Bode plots
% @author: Mirco Meiners (HSB)
% Input signal: DF_IN1
% Output signal: DF_IN2
clear;

%% Define Red Pitaya as TCP client object
IP = '192.168.111.181';  % IP of your Red Pitaya ...
port = 5000;
RP = tcpclient(IP, port);

%% Open connection to Red Pitaya
RP.ByteOrder = 'big-endian';
configureTerminator(RP, 'CR/LF');
flush(RP);

%% Generate continous signal
dec = 16;
trig_lvl = 0.5;
gain = 'LV';
trig_dly = 8192;
acq_trig = 'CH1_PE';
func = 'TRIANGLE';  % Signalform (SINE, SQUARE, TRIANGLE)
ampl = 0.5;  % Amplitude (-1 V ... 1 V)
freq = 1000;  % Frequenz (0 Hz ... 62.5e6 Hz)
offset = 0.5;  % Offset (-1 V ... 1 V)
t = linspace(0, 2.097e-3, 16384);  % Skalierung der Zeitachse (Sampling)


% Send SCPI command to Red Pitaya to turn ON generator
writeline(RP,'GEN:RST');  % Reset Generator
writeline(RP,append('SOUR1:FUNC',' ',func));  % Set function of output signal
writeline(RP,append('SOUR1:VOLT',' ',num2str(ampl)));  % Set amplitude
writeline(RP,append('SOUR1:VOLT:OFFS',' ',num2str(offset)));  % Set offset
writeline(RP,append('SOUR1:FREQ:FIX',' ',num2str(freq)));  % Set frequency
writeline(RP,'OUTPUT1:STATE ON');  % Turn on output OUT1
    
writeline(RP,'SOUR1:TRig:INT');  % Generate trigger
    
pause(1);

% ACQUISITION
writeline(RP,'ACQ:RST ');  % Input reset
% writeline(RP,'ACQ:DATA:FORMAT ASCII')
% writeline(RP,'ACQ:DATA:Units VOLTS')
writeline(RP,append('ACQ:DEC:Factor',' ',num2str(dec)));  % Decimation (1, 8, 16, 64, 1024, 8192)
writeline(RP,append('ACQ:TRig:LEV',' ',num2str(trig_lvl)));  % Trigger level

% Set trigger delay to 0 samples
% 0 samples delay sets trigger to the center of the buffer
% Signal on your graph will have the trigger in the center (symmetrical)
% Samples from left to the center are samples before trigger
% Samples from center to the right are samples after trigger
writeline(RP,append('ACQ:TRig:DLY',' ',num2str(trig_dly)));  % Delay

% Start & Trigger
% Trigger source setting must be after ACQ:START
% Set trigger to source 1 positive edge
writeline(RP,'ACQ:START')  % Start der Messung

% Sets gain to LV/HV (should be the same as jumpers)
% writeline(RP,append('ACQ:SOUR1:GAIN',' ',num2str(gain)));  
% writeline(RP,append('ACQ:SOUR2:GAIN',' ',num2str(gain)));  

% After acquisition is started some time delay is needed in order to acquire fresh samples in the buffer
pause(1);
% Here we have used time delay of one second, but you can calculate the exact value by taking into account buffer
% length and sampling rate

writeline(RP,'ACQ:TRig NOW')  % Trigger manually

% This loop is not needed as we triggered the acquisition manually
%while 1
%    trig_rsp = writeread(RP,'ACQ:TRig:STAT?')
%    if strcmp('TD', trig_rsp(1:2))  % Read only TD
%        break;
%    end
%end

% wait for fill adc buffer
while 1
    fill_state = writeread(RP,'ACQ:TRig:FILL?')
    if strcmp('1', fill_state(1:1))
        break;
    end
end

for n = 1:2
    % Input IN1
    IN1 = writeread(RP,'ACQ:SOUR1:DATA?');
    IN1_num = str2num(IN1(2:length(IN1)-1));
    DF_IN1(:,n) = IN1_num';

    % Input IN2
    IN2 = writeread(RP,'ACQ:SOUR2:DATA?');
    IN2_num = str2num(IN2(2:length(IN2)-1));
    DF_IN2(:,n) = IN2_num';
end

% Turn off generator OUT1
writeline(RP,'OUTPUT1:STATE OFF');
    
%% Close connection to Red Pitaya
clear RP;

%% Save data as mat file
% save('./data/IN_D1.mat', 'DF_IN1', 'DF_IN2');
save('./data/IN1_D1_ELIE1.mat', 'DF_IN1');
save('./data/IN2_D1_ELIE1.mat', 'DF_IN2');

%% Save data as csv file
% Not recommended, old version
% csvwrite('./data/IN1_D1.csv', 'DF_IN1');
% csvwrite('./data/IN2_D1.csv', 'DF_IN2'); 

% Better cross-platform compatibility, new version
% writematrix('DF_IN1', './data/IN1_D1.mat.csv');
% writematrix('DF_IN2', './data/IN2_D1.mat.csv');

%% Save data as parquet file
% parquet data is of type table, no matrix operations
% parquetwrite('data/IN1_D1.mat.parquet', array2table(DF_IN1));
% parquetwrite('data/IN2_D1.mat.parquet', array2table(DF_IN2));

%% Save data as excel sheet
% data is table data, no matrix operations
% writematrix(DF_IN1, './data/IN1_D1.mat.xlsx');
% writematrix(DF_IN1, './data/IN2_D1.mat.xlsx');
% writematrix(DF_IN1, './data/IN_D1.mat.xlsx', 'Sheet', 1);
% writematrix(DF_IN2, './data/IN_D1.mat.xlsx', 'Sheet', 2);
